doc/developer/platform/architecture-storage.md

```diff
@@ -107,25 +107,142 @@ All commands are durably persisted and their effect survives crashes and reboots
+
+    As suggested, this may include things like the schema registry that aren't conventional inputs, but nonetheless are external sources that influence the output we will produce.
+
+2. Identify the "state" that a streaming implementation of the intended functionality would need to maintain.
+
+    For example, if we are matching up transaction metadata with the data it references, we may need to maintain some outstanding metadata records and some outstanding data records, until all in a transaction match.
+    If we intend to hold back transactions until all strictly prior transactions have completed, we may need to hold on even to some completed transactions, awaiting the completion of the prior work.
+
+3. Create `persist` collections for the output and the identified "state".
+
+
+With this prep work done, the assembly instructions are:
+
+1. To start, determine `t` the `min` write frontier in `persist` for all outputs and all state.
+
+2. From time `t` onward, rehydrate the "state" associated with the source pipeline.
+
+3. Restart each raw source from the addresses found in the persisted reclocking collection at `t`.
+
+4. From time `t` onward, continually produce reclocked messages: messages from the raw source whose addresses are translated to time using persist's reclocking information.
+
+    It is important that this be based on reclocking information we hear back from persist, rather than what we optimistically proposed at persist.
+    This ensures that the right thing happens in the case there are two contending implementations.
+
+    Each raw source should produce exactly its outputs at times greater or equal to `t`.
+
+5. From time `t` onward, apply the source pipeline to produce at each time:
+
+    a. Changes to push at persist to update the output.
+
+    b. Changes to push at persist to update the state.
+
+These are the whole instructions.
+Clearly Step 5. is not fully specified.
+The property it must have is that two independent pipelines should produce the same output and state updates.
+One way to make this happen is to behave as if each message becomes available at its timestamp, and at each time we "immediately" retire all complete work.
+For example, the first timestamp for which we have both transaction metadata and all supporting data messages, we would output those data messages.
+Alternately, the first timestamp for which we have both transaction metadata and all supporting data messages, *and no incomplete prior transactions*, we would output those data messages.
+The specific choice of what to do depends on our goals, but one attempts to do it as if executing each timestamp one at a time in sequence, and doing all the work that is possible at each timestamp.
+
+### Outcomes
+
+The reason we follow the recipe is because many good things happen automatically.
+
+1. *Definiteness.*
+
+    We get definiteness from `persist`.
+    I don't think we are able to screw this up.
```

Comment:
This is a bit repetitive with the intro of these properties; not sure if intentional!

Reply:
It is! I just re-read it, and happy with it. The text is more substantial here, and is meant to present the rationale for why each of the properties might be satisfied, a discussion we cannot have when they are first introduced.

Reply:
Same! I just meant property (1) which had the same "I don't think we can screw this up" text in both places. The new revision has that text only down here, which seems great!